In [2]:
import pandas as pd
import numpy as np

In [3]:
df = pd.read_csv("heart.csv")
df.head()

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  MaxHR  \
0   40   M           ATA        140          289          0     Normal    172   
1   49   F           NAP        160          180          0     Normal    156   
2   37   M           ATA        130          283          0         ST     98   
3   48   F           ASY        138          214          0     Normal    108   
4   54   M           NAP        150          195          0     Normal    122   

  ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0              N      0.0       Up             0  
1              N      1.0     Flat             1  
2              N      0.0       Up             0  
3              Y      1.5     Flat             1  
4              N      0.0       Up             0

In [4]:
print(df.info())
print(df.describe())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 918 entries, 0 to 917
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Age             918 non-null    int64  
 1   Sex             918 non-null    object 
 2   ChestPainType   918 non-null    object 
 3   RestingBP       918 non-null    int64  
 4   Cholesterol     918 non-null    int64  
 5   FastingBS       918 non-null    int64  
 6   RestingECG      918 non-null    object 
 7   MaxHR           918 non-null    int64  
 8   ExerciseAngina  918 non-null    object 
 9   Oldpeak         918 non-null    float64
 10  ST_Slope        918 non-null    object 
 11  HeartDisease    918 non-null    int64  
dtypes: float64(1), int64(6), object(5)
memory usage: 86.2+ KB
None
              Age   RestingBP  Cholesterol   FastingBS       MaxHR  \
count  918.000000  918.000000   918.000000  918.000000  918.000000   
mean    53.510893  132.396514   198.799564    0.233115  1

In [5]:
df.duplicated().sum()

np.int64(0)

In [6]:
df.dtypes

Age                 int64
Sex                object
ChestPainType      object
RestingBP           int64
Cholesterol         int64
FastingBS           int64
RestingECG         object
MaxHR               int64
ExerciseAngina     object
Oldpeak           float64
ST_Slope           object
HeartDisease        int64
dtype: object

In [7]:
df["Sex"].unique()

array(['M', 'F'], dtype=object)

In [8]:
df["Sex"] = df["Sex"].map({"M": 1, "F": 0})
df.head()

Age  Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
0   40    1           ATA        140          289          0     Normal   
1   49    0           NAP        160          180          0     Normal   
2   37    1           ATA        130          283          0         ST   
3   48    0           ASY        138          214          0     Normal   
4   54    1           NAP        150          195          0     Normal   

   MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0    172              N      0.0       Up             0  
1    156              N      1.0     Flat             1  
2     98              N      0.0       Up             0  
3    108              Y      1.5     Flat             1  
4    122              N      0.0       Up             0

In [9]:
df["ChestPainType"].unique()

array(['ATA', 'NAP', 'ASY', 'TA'], dtype=object)

In [10]:
df = pd.get_dummies(df, columns=["ChestPainType"])
df.head()

Age  Sex  RestingBP  Cholesterol  FastingBS RestingECG  MaxHR  \
0   40    1        140          289          0     Normal    172   
1   49    0        160          180          0     Normal    156   
2   37    1        130          283          0         ST     98   
3   48    0        138          214          0     Normal    108   
4   54    1        150          195          0     Normal    122   

  ExerciseAngina  Oldpeak ST_Slope  HeartDisease  ChestPainType_ASY  \
0              N      0.0       Up             0              False   
1              N      1.0     Flat             1              False   
2              N      0.0       Up             0              False   
3              Y      1.5     Flat             1               True   
4              N      0.0       Up             0              False   

   ChestPainType_ATA  ChestPainType_NAP  ChestPainType_TA  
0               True              False             False  
1              False               True             False  
2               True              False             False  
3              False              False             False  
4              False               True             False

In [11]:
df["ChestPainType_TA"] = df["ChestPainType_TA"].astype(int)
df["ChestPainType_NAP"] = df["ChestPainType_NAP"].astype(int)
df["ChestPainType_ASY"] = df["ChestPainType_ASY"].astype(int)
df["ChestPainType_ATA"] = df["ChestPainType_ATA"].astype(int)
df.head()

Age  Sex  RestingBP  Cholesterol  FastingBS RestingECG  MaxHR  \
0   40    1        140          289          0     Normal    172   
1   49    0        160          180          0     Normal    156   
2   37    1        130          283          0         ST     98   
3   48    0        138          214          0     Normal    108   
4   54    1        150          195          0     Normal    122   

  ExerciseAngina  Oldpeak ST_Slope  HeartDisease  ChestPainType_ASY  \
0              N      0.0       Up             0                  0   
1              N      1.0     Flat             1                  0   
2              N      0.0       Up             0                  0   
3              Y      1.5     Flat             1                  1   
4              N      0.0       Up             0                  0   

   ChestPainType_ATA  ChestPainType_NAP  ChestPainType_TA  
0                  1                  0                 0  
1                  0                  1                 0  
2                  1                  0                 0  
3                  0                  0                 0  
4                  0                  1                 0

In [12]:
df["RestingECG"].unique()

array(['Normal', 'ST', 'LVH'], dtype=object)

In [13]:
df = pd.get_dummies(df, columns=["RestingECG"])

In [14]:
df.head()

Age  Sex  RestingBP  Cholesterol  FastingBS  MaxHR ExerciseAngina  Oldpeak  \
0   40    1        140          289          0    172              N      0.0   
1   49    0        160          180          0    156              N      1.0   
2   37    1        130          283          0     98              N      0.0   
3   48    0        138          214          0    108              Y      1.5   
4   54    1        150          195          0    122              N      0.0   

  ST_Slope  HeartDisease  ChestPainType_ASY  ChestPainType_ATA  \
0       Up             0                  0                  1   
1     Flat             1                  0                  0   
2       Up             0                  0                  1   
3     Flat             1                  1                  0   
4       Up             0                  0                  0   

   ChestPainType_NAP  ChestPainType_TA  RestingECG_LVH  RestingECG_Normal  \
0                  0                 0           False               True   
1                  1                 0           False               True   
2                  0                 0           False              False   
3                  0                 0           False               True   
4                  1                 0           False               True   

   RestingECG_ST  
0          False  
1          False  
2           True  
3          False  
4          False

In [15]:
df["RestingECG_LVH"] = df["RestingECG_LVH"].astype(int)
df["RestingECG_Normal"] = df["RestingECG_Normal"].astype(int)
df["RestingECG_ST"] = df["RestingECG_ST"].astype(int)
df.head()

Age  Sex  RestingBP  Cholesterol  FastingBS  MaxHR ExerciseAngina  Oldpeak  \
0   40    1        140          289          0    172              N      0.0   
1   49    0        160          180          0    156              N      1.0   
2   37    1        130          283          0     98              N      0.0   
3   48    0        138          214          0    108              Y      1.5   
4   54    1        150          195          0    122              N      0.0   

  ST_Slope  HeartDisease  ChestPainType_ASY  ChestPainType_ATA  \
0       Up             0                  0                  1   
1     Flat             1                  0                  0   
2       Up             0                  0                  1   
3     Flat             1                  1                  0   
4       Up             0                  0                  0   

   ChestPainType_NAP  ChestPainType_TA  RestingECG_LVH  RestingECG_Normal  \
0                  0                 0               0                  1   
1                  1                 0               0                  1   
2                  0                 0               0                  0   
3                  0                 0               0                  1   
4                  1                 0               0                  1   

   RestingECG_ST  
0              0  
1              0  
2              1  
3              0  
4              0

In [16]:
df.dtypes

Age                    int64
Sex                    int64
RestingBP              int64
Cholesterol            int64
FastingBS              int64
MaxHR                  int64
ExerciseAngina        object
Oldpeak              float64
ST_Slope              object
HeartDisease           int64
ChestPainType_ASY      int64
ChestPainType_ATA      int64
ChestPainType_NAP      int64
ChestPainType_TA       int64
RestingECG_LVH         int64
RestingECG_Normal      int64
RestingECG_ST          int64
dtype: object

In [17]:
df["ExerciseAngina"].unique()

array(['N', 'Y'], dtype=object)

In [18]:
df["ExerciseAngina"] = df["ExerciseAngina"].map({"N": 0, "Y": 1})
df.head()

Age  Sex  RestingBP  Cholesterol  FastingBS  MaxHR  ExerciseAngina  \
0   40    1        140          289          0    172               0   
1   49    0        160          180          0    156               0   
2   37    1        130          283          0     98               0   
3   48    0        138          214          0    108               1   
4   54    1        150          195          0    122               0   

   Oldpeak ST_Slope  HeartDisease  ChestPainType_ASY  ChestPainType_ATA  \
0      0.0       Up             0                  0                  1   
1      1.0     Flat             1                  0                  0   
2      0.0       Up             0                  0                  1   
3      1.5     Flat             1                  1                  0   
4      0.0       Up             0                  0                  0   

   ChestPainType_NAP  ChestPainType_TA  RestingECG_LVH  RestingECG_Normal  \
0                  0                 0               0                  1   
1                  1                 0               0                  1   
2                  0                 0               0                  0   
3                  0                 0               0                  1   
4                  1                 0               0                  1   

   RestingECG_ST  
0              0  
1              0  
2              1  
3              0  
4              0

In [19]:
df["ST_Slope"].unique()

array(['Up', 'Flat', 'Down'], dtype=object)

In [20]:
df = pd.get_dummies(df, columns=["ST_Slope"])
df.head()

Age  Sex  RestingBP  Cholesterol  FastingBS  MaxHR  ExerciseAngina  \
0   40    1        140          289          0    172               0   
1   49    0        160          180          0    156               0   
2   37    1        130          283          0     98               0   
3   48    0        138          214          0    108               1   
4   54    1        150          195          0    122               0   

   Oldpeak  HeartDisease  ChestPainType_ASY  ChestPainType_ATA  \
0      0.0             0                  0                  1   
1      1.0             1                  0                  0   
2      0.0             0                  0                  1   
3      1.5             1                  1                  0   
4      0.0             0                  0                  0   

   ChestPainType_NAP  ChestPainType_TA  RestingECG_LVH  RestingECG_Normal  \
0                  0                 0               0                  1   
1                  1                 0               0                  1   
2                  0                 0               0                  0   
3                  0                 0               0                  1   
4                  1                 0               0                  1   

   RestingECG_ST  ST_Slope_Down  ST_Slope_Flat  ST_Slope_Up  
0              0          False          False         True  
1              0          False           True        False  
2              1          False          False         True  
3              0          False           True        False  
4              0          False          False         True

In [21]:
df["ST_Slope_Down"] = df["ST_Slope_Down"].astype(int)
df["ST_Slope_Flat"] = df["ST_Slope_Flat"].astype(int)
df["ST_Slope_Up"] = df["ST_Slope_Up"].astype(int)
df.head()

Age  Sex  RestingBP  Cholesterol  FastingBS  MaxHR  ExerciseAngina  \
0   40    1        140          289          0    172               0   
1   49    0        160          180          0    156               0   
2   37    1        130          283          0     98               0   
3   48    0        138          214          0    108               1   
4   54    1        150          195          0    122               0   

   Oldpeak  HeartDisease  ChestPainType_ASY  ChestPainType_ATA  \
0      0.0             0                  0                  1   
1      1.0             1                  0                  0   
2      0.0             0                  0                  1   
3      1.5             1                  1                  0   
4      0.0             0                  0                  0   

   ChestPainType_NAP  ChestPainType_TA  RestingECG_LVH  RestingECG_Normal  \
0                  0                 0               0                  1   
1                  1                 0               0                  1   
2                  0                 0               0                  0   
3                  0                 0               0                  1   
4                  1                 0               0                  1   

   RestingECG_ST  ST_Slope_Down  ST_Slope_Flat  ST_Slope_Up  
0              0              0              0            1  
1              0              0              1            0  
2              1              0              0            1  
3              0              0              1            0  
4              0              0              0            1

In [22]:
df.dtypes

Age                    int64
Sex                    int64
RestingBP              int64
Cholesterol            int64
FastingBS              int64
MaxHR                  int64
ExerciseAngina         int64
Oldpeak              float64
HeartDisease           int64
ChestPainType_ASY      int64
ChestPainType_ATA      int64
ChestPainType_NAP      int64
ChestPainType_TA       int64
RestingECG_LVH         int64
RestingECG_Normal      int64
RestingECG_ST          int64
ST_Slope_Down          int64
ST_Slope_Flat          int64
ST_Slope_Up            int64
dtype: object

In [23]:
x = df.drop('HeartDisease', axis=1)
y = df['HeartDisease']

In [25]:
from sklearn.model_selection import train_test_split

In [28]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)

In [29]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report

In [58]:
model = LogisticRegression(max_iter=10000)
model.fit(x_train, y_train)
y_pred = model.predict(x_test)


In [59]:
accuracy = accuracy_score(y_test, y_pred)
accuracy_percentage = accuracy * 100

print(f"Accuracy: {accuracy_percentage:.2f}%")

Accuracy: 85.33%


In [60]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.80      0.87      0.83        77
           1       0.90      0.84      0.87       107

    accuracy                           0.85       184
   macro avg       0.85      0.86      0.85       184
weighted avg       0.86      0.85      0.85       184



# Understanding the classification report for my heart disease model
### Here’s what the main metrics mean and how my model performed:

Support is how many actual samples there are for each class in the test data:

- Class 0 (no heart disease): 77 samples

- Class 1 (heart disease): 107 samples

Precision tells me, out of all the times the model predicted a class, how often it was correct:

- For class 0, the model was right 82% of the time it predicted “no disease.”

- For class 1, it was right 90% of the time it predicted “heart disease.”

Recall tells me, out of all actual cases of a class, how many the model correctly found:

- For class 0, the model correctly identified 87% of all “no disease” cases.

- For class 1, it found 86% of all “heart disease” cases.

F1-score is a balance between precision and recall (higher is better):

- 0.84 for class 0

- 0.88 for class 1

Accuracy is the overall percentage of correct predictions: about 86%.

Macro average is the simple average of the precision, recall, and F1-score for both classes — treating them equally.

Weighted average takes the number of samples per class into account.

Summary:
- The model is slightly better at predicting heart disease (class 1), which is great because catching those cases is important. Recall values show the model doesn’t miss too many cases in either class. Overall, the model performs well!



In [61]:
new_data = np.array([[45, 0, 150, 230, 0, 165, 0, 0.5, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0]])
predicted = model.predict(new_data)
predicted[0]

c:\Users\taham\OneDrive\Desktop\coding\.venv\Lib\site-packages\sklearn\utils\validation.py:2749: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(


np.int64(0)

In [62]:
predicted_proba = model.predict_proba(new_data)
predicted_proba

c:\Users\taham\OneDrive\Desktop\coding\.venv\Lib\site-packages\sklearn\utils\validation.py:2749: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(


array([[0.94615589, 0.05384411]])